text/3735-field-projections.md

```diff
@@ -0,0 +1,1555 @@
+}
+
+pub mod bar {
+    // Error: field `bar` of type `Foo` is private
+    type FooBar = field_of!(Foo, bar);
+}
+```
+
+#### `#[projecting]` Attribute
+
+The `#[projecting]` attribute can be put on a struct or union declaration. It requires that the
+type is `#[repr(transparent)]` and that there is a unique field that has non-zero size in general.
+It results in fields of the single non-zero sized type being considered fields of the outer type.
+
+So for example:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container<T> {
+    inner: T,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
+
+Now `Container<Foo>` has a [field type] associated with `bar` implementing `Field` with:
+- `Base = Container<Foo>`
+- `Type = Container<i32>`
+- `OFFSET = offset_of!(Foo, bar)`
+
+The same is true for the non-generic case:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container {
+    inner: Foo,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
+
+Here are some error examples:
+
+```rust
+// ERROR: missing `#[repr(transparent)]`
+#[projecting]
+pub struct Container<T> {
+    inner: T,
+}
+
+// ERROR: no field to project onto found, the struct has no fields
+#[projecting]
+#[repr(transparent)]
+pub struct Container {}
+
+// ERROR: no field to project onto found, all fields are always zero-sized
+#[projecting]
+#[repr(transparent)]
+pub struct Container<T> {
+    phantom1: PhantomData<fn() -> T>,
+    phantom2: PhantomData<fn(T)>,
+}
+```
+
+### Field Projection Operator
+
+The field projection operator `->` has the following syntax:
+
+> **<sup>Syntax</sup>**\
+> _ProjectionExpression_ :\
+> &nbsp;&nbsp; [_Expression_] `->` _ProjectionMember_
+>
+> _ProjectionMember_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [IDENTIFIER]
+> &nbsp;&nbsp; | [TUPLE_INDEX]
+
+[IDENTIFIER]: https://doc.rust-lang.org/reference/identifiers.html
+[_Expression_]: https://doc.rust-lang.org/reference/expressions.html
+[TUPLE_INDEX]: https://doc.rust-lang.org/reference/tokens.html#tuple-index
+
+#### Desugaring
+
+```rust
+struct T {
+    field: F,
+}
+
+let t: C<T> = /* ... */;
+let _ = t->field;
+
+// becomes
+
+let _ = Project::<field_of!(<C<T> as Projectable>::Inner, field)>::project(t);
+```
+
+The `C<T>` in the `C<T> as Projectable` comes from a type inference variable over the expression
+`t`.
+
+
+#### `Projectable` & `Project` Traits
+
+Added to `core::ops`:
+
+```rust
+pub trait Projectable: Sized {
+    type Inner: ?Sized;
+}
+
+pub trait Project<F>: Projectable
+where
+    F: Field<Base = Self::Inner>,
+{
+    type Output;
+
+    fn project(self) -> Self::Output;
+}
+```
+
+## Stdlib Field Projections
+
+All examples from the guide-level explanation work when the standard library is extended with the
+implementations detailed below.
+
+The following pointer types get an implementation for `Projectable` with `Inner = T`. They support
+projections for any field and perform the obvious offset operation.
+
+- `&mut T`
+- `&T`
+- `*mut T`
+- `*const T`
+- `NonNull<T>`
```

Comment:
Are `cell::Ref<T>` and `cell::RefMut<T>` supposed to be in this list as well? They are mentioned in the Guild-Level Explanation but not here.

`MappedMutexGuard<T>`, `MappedRwLockReadGuard<T>`, `MappedRwLockWriteGuard<T>` as well?

Reply:
yes, those are some great suggestions! (:eyes: didn't know about those!)